### ISSUE TITLE
prepend arguments of type struct with "struct"'

### ISSUE THREAD

**Issue by teto** [gccxml/pygccxml#63]
the decl_string of for instance "struct sysinfo" should be sysinfo and not just "sysinfo".

For instance matching the `sysinfo` syscall and printing its first parameter decl_string returns
" ::sysinfo *"

When printing the result of the matcher.find, it gives

```
resultats= [<pygccxml.declarations.free_calldef.free_function_t object at 0x7f83a8262eb8>]
resultat: extern int sysinfo(sysinfo * __info) [free function]
```

I would expect 
`resultat: extern int sysinfo(struct sysinfo * __info) [free function]`

**Comment by iMichka** [comment]
I had a look at the sys/sysinfo.h file.

We are talking about the following code, right ?
```
struct sysinfo {
 ...
};

extern int sysinfo (struct sysinfo *__info);
```

When writing:
 ```
struct sysinfo {
 ...
};
```
, this means the ```sysinfo``` identifier is defined within the struct name space., see: http://stackoverflow.com/questions/1675351/typedef-struct-vs-struct-definitions/1675446#1675446

Now, the question is, do I need the ```struct``` keyword in the parameter definition ? It seems that this keyword is "optional", at least as long as there is no ambiguity: http://stackoverflow.com/a/8423005. The ```struct``` keyword  is coming from C, and could be skipped. In ```sys/sysinfo.h```, there is no ambiguity so the struct keyword could be removed, but I guess the file is C and C++ compatible so they are keeping it around.

Compiling the code with or without ```struct``` keyword in the parameters works for me with Clang and GCC 6 on OS X. So as this is optional, I am not sure I want to add the further logic for this.

You could argue that using ```clang -Xclang -ast-dump -fsyntax-only``` would output something like this:
```
|-CXXRecordDecl 0x104023600 <test_struct_type.hpp:15:1, col:17> col:8 referenced struct sysinfo definition
| `-CXXRecordDecl 0x104023710 <col:1, col:8> col:8 implicit struct sysinfo
`-FunctionDecl 0x1040238a0 <line:19:1, col:36> col:12 sysinfo 'int (struct sysinfo *)' extern
  `-ParmVarDecl 0x1040237d8 <col:21, col:30> col:30 __info 'struct sysinfo *'
```

Looking at the xml file produced by CastXML, this information about the struct keyword is not provided. If we want to add that information, we need first to patch CastXML to output that information. Then we need to read it into pygccxml and display it.

It really depends on what you want to do later with that information, and if it is important to you.
The declaration tree seems consistent to me. If ```sysinfo_decl``` is your ```free_function_t``` declaration called ```sysinfo```, you will have:
```
print(declarations.remove_pointer(sysinfo_decl.arguments[0].decl_type).declaration.name)
> sysinfo
print(type(declarations.remove_pointer(sysinfo_decl.arguments[0].decl_type).declaration))
> <class 'pygccxml.declarations.class_declaration.class_t'>
```

If you really need that struct there, another solution would be to always add it to the output. The clang AST dump example I gave is always displaying it, even if ```struct``` is omitted.

**Comment by teto** [comment]
Yes we are talking about that code :)
I am writing a C program so the "struct" keyword is mandatory (Even in C++, I would most likely prefer to add the "struct" for the same reason the linux kernel bans typedef: readability). with gcc 5, my program won't compile without the "struct" keyword.
If I call declarations.remove_pointer on the "sysinfo *" parameter, I obtain an object of a type "class 'pygccxml.declarations.cpptypes.declarated_t'>" , can I assume "declarated_t" imply struct  ? in which case it would be an easy fix ?

**Comment by praetorian20** [comment]
@iMichka Does CastXML not have `struct` in the output even in C mode (`--castxml-cc-gnu-c`)? As @teto notes, you need it there for a C program, while it's redundant for a C++ program. If it's absent in C mode, then I'd say that's a CastXML bug.

I guess this raises another question, does pygccxml support C compatible output? I've only ever used it for C++.

**Comment by iMichka** [comment]
@teto I was not completely aware that you were parsing C code. I think you mentioned it somewhere else but I forgot about it.

As @praetorian20 points out, I am not sure pygccxml completely works with C code. All the tests have been running with C++ mode since I overtook the project. Also, I myself never used it for anything else than C++. Of course, in principle most things will just be working as C++ and C are very similar. But small discrepancies like this one do exist.

Also, the CastXML command line is explicitly set to C++ here (with ```-x c++``` for clang):
https://github.com/gccxml/pygccxml/blob/develop/pygccxml/parser/source_reader.py#L162-L163

@praetorian20 I gave it a try with ```--castxml-cc-gnu-c```. It does not change the output of the xml file. No "struct" keyword is appearing (Tested with gcc or clang, and also by removing ```-x c++```).

@teto ```declarated_t``` is a class that wraps a type and a declaration. You can call .declaration on it to access the actual declaration. If you want to check if it is a struct, you can use the following check:
```decl.class_type == class_declaration.CLASS_TYPES.STRUCT```. I may add an ```is_struct```method later, which seems to be missing in pygccxml.

In conclusion:
- pygccxml does not support C code for the moment, or at least it is untested.
- supporting C it out of scope of the project, but I would be willing to review pull requests
- CastXML may have bug because it should know about the struct keyword. Not sure how the xml output should look in that case though. @bradking may have more insights on that point.

If we want to support C correctly:
- The test suite should a least be run with a C compiler on Travis (I can take care of that part)
- I just rapidly ran the examples in C mode, they are failing as there are some syntax errors. One would first need to clean them up.
- Special-casing code for C may add complexity. It may also alter performance, so we need to be careful.
- If only the struct thingy is wished, I can add a flag that you can set as option, to always add it, be it in C++ or C. This may be the faster way to achieve @teto's goal. But if more things are to come for C support, this will not help on a long term.

**Comment by teto** [comment]
I haven't met any problem apart from that and the variadic support ("va_list") arguments which I was able to hack around with
```
s = str(arg.decl_type)
if s.startswith("?unknown?"):
                    # handles variables list
                    log.debug("UNKNOWN => variadic")
                    s = "va_list " + decl.arguments[-1].name
```
I suppose it's not 100% reliable but in my case it works pretty well.

Having a flag to enforce or not the "struct" keyword sounds like a good idea to comply with user preference, it must be overidden to true in C mode.
For now, I've added special cases for "sysinfo" and a few other functions so there is no hurry even if I would prefer to be the code generation fully automatic via pygccxml.

**Comment by iMichka** [comment]
I dug a little bit though the c++ documentation. This thing is called an "elaborated type specifier" if I get it right (http://en.cppreference.com/w/cpp/language/elaborated_type_specifier).

If I want to implement that correctly, I would need to add it for class, struct, union and enum. I never wrote pure C code, so would those 4 also make sense there ? If I add a flag to enable them, I will probably need to do it for all of them. Or should there be 4 separate flags, for each of the elaborated type specifiers ?

**Comment by praetorian20** [comment]
I would vote for a single flag, but I'm not sure adding a flag is the right fix to begin with. What about the case where someone defines a `struct` like this (quite common in C)

```
typedef struct foo {} foo;
```

In that case, `foo` works even in C and `struct foo` isn't needed. IMHO, this is something that should be fixed in CastXML. If the `struct` is required for code to be syntactically valid, it should be present in the XML output.

**Comment by iMichka** [comment]
Yes, let's see first if we can get that information in CastXML first.

I also do not really like the flag thing. It means maintaining different tests for different flags, which is expensive. And also you would only have an on/off switch for all elaborated type specifiers, but you may want to track only those which were really written in your c++ initial source.

There seems a way to get that information here: http://clang.llvm.org/doxygen/classclang_1_1ElaboratedType.html
I need to wrap my head around on how to add that to CastXML, without breaking backward compatibility.

**Comment by iMichka** [comment]
Just to let you know that I am still working on a solution, and I am making good progress. I will submit a patch to CastXML next week, and the pygccxml is almost ready too.

**Comment by teto** [comment]
Fantastic, I will resume my work on it when it's ready then. Thanks for the heads-up !

**Comment by iMichka** [comment]
I just opened the patch for CastXML here: https://github.com/CastXML/CastXML/pull/77

Once the patch is merged to CastXML, I will add support for this feature in pygccxml. I am almost done with writing the tests, so this could go fast now.

**Comment by iMichka** [comment]
Hi

So finally the patch I made would have worked but did not take into account all the cases. The situation is a little bit more complicated than thought. You can read about it in the CastXML bug.

I will start working on implementing a new XML file format for CastXML, which will allow to implement the elaborated types feature. We need to do this because the changes would break the old GCCXML format. So we are working on this; but it will take even more time than thought.

BTW, I released 1.8.1 which contains the fix for the function pointers.

**Comment by teto** [comment]
I subscribed to the other post so I see the progress.
I've resumed working on my related work but I have some workarounds for the current problem so that's ok.

As a side question, is it possible to copy a structure declaration with pygccxml. For instance the libc defines structures that are not exported and I would like to be able to copy those declarations.

once again thanks for your work, the required changes seem to be quite sthg :)

**Comment by iMichka** [comment]
FYI: Patch has been merged in CastXML; the elaborated types information can now be generated.
I already implemented code to generate the files with the new format, using pygccxml.

The next step will be to fully support this in pygccxml. I will start with this during the week.

There is also a bunch of CI that need to be tweaked to have the tests running with the old and the new format. It's not really blocking this but I do want to have the tests running the newest version. For ref.: #74.

**Comment by teto** [comment]
cool stuff ! let me know when there is a branch you want me to test. Thanks for your dedication :+1:

**Comment by iMichka** [comment]
@teto I pushed the last changes, you can test this now on the develop branch. You will need a recent castxml, and set `castxml_epic_version=1` as flag in the configuration for pygccxml.

There is now a new `elaborated_t` type, which works similarly to the others qualifiers types like `const_t` etc...

I will need a few weeks to finish the 1.9.0 release, I have some details to take care of, but it would be great if you could give this a try.

**Comment by teto** [comment]
Thanks for the update. I just installed castxml-dev, will try to test teh changes this week. Any example I could/should look at for elaborated_t use ?

**Comment by teto** [comment]
I couldn't find a "castxml_epic_version" string in castxml code I believe it's "--castxml-output=1" instead ?

**Comment by teto** [comment]
My bad I see that was a pygccxml flag.

I wonder if that check is necessary as castxml will throw an error anyway
```
        if self.__config.castxml_epic_version is not None:
            if self.__config.castxml_epic_version != 1:
                raise RuntimeError(
                    "The CastXML epic version can only be 1, "
                    "but it was " + str(self.__config.castxml_epic_version))
```

**Comment by iMichka** [comment]
> Any example I could/should look at for elaborated_t use ?

Nothing fancy here, it works as he other `compound_t` objects. I just added a small example to document this new behaviour: http://pygccxml.readthedocs.io/en/develop/examples/elaborated/example.html

> I couldn't find a "castxml_epic_version" string in castxml code I believe it's "--castxml-output=1" instead ?

As you found out, `castxml_epic_version` is a pygccxml flag.

> I wonder if that check is necessary as castxml will throw an error anyway

That check is only there to make sure that if somebody provides that flag, it is set to 1. Nothing more. If castxml is an older version and the flag is provided, then castxml will throw an error an tell the user that this is not working. I should maybe add a try/catch more around the castxml call, and throw my own exception telling the user that he can't do this with older castxml versions, but this is probably overkill.

**Comment by iMichka** [comment]
I'm closing this issue, as the 1.9.0 release is out since some time now. You can still comment here or open a new issue if you have any problems with this new feature. Feedback is welcome. Thanks for your patience.
